main.go

```diff
@@ -608,3 +615,84 @@ func IndexCIEDE2000(pal color.Palette, orig color.Color) int {
 	}
 	return ret
 }
+
+// Color can convert itself to alpha-premultiplied 16-bits per channel RGBA.
+// The conversion may be lossy.
+type Color interface {
+	RGBA() (r, g, b, a uint32)
+}
+
+// Palette is a palette of colors.
+type Palette []Color
+
+// Find the brightest color
+func GreatestColor(r int, g int, b int) int {
+	if r > g && r > b {
+		return 1
+	}
+	if g > r && g > b {
+		return 2
+	}
+	if b > r && b > g {
+		return 3
+	}
+
+	return 0
+}
+
+// Index returns the index of the palette color closest to c in Euclidean
+// R,G,B,A space. Strongest color multiplied by the treshold value.
+func IndexMult(p color.Palette, c Color, thresh uint32) int {
+	cr, cg, cb, ca := c.RGBA()
+	// Is this color visibly red, green, or blue?
+	brightest := GreatestColor(int(cr), int(cg), int(cb))
+	ret, bestSum := 0, uint32(1<<32-1)
+	for i, v := range p {
+		vr, vg, vb, va := v.RGBA()
+		rr := sqDiff(cr, vr)
+		gg := sqDiff(cg, vg)
+		bb := sqDiff(cb, vb)
+		aa := ca - va
+		if aa != 0 {
+			aa = 0
+		}
+		if brightest == 1 {
+			rr *= thresh
+		}
+		if brightest == 2 {
+			gg *= thresh
+		}
+		if brightest == 3 {
+			bb *= thresh
+		}
+		sum := rr + gg + bb
+		if sum < bestSum {
+			if sum == 0 {
+				return i
+			}
+			ret, bestSum = i, sum
+		}
+	}
+	return ret
+}
+
+// sqDiff returns the squared-difference of x and y, shifted by 2 so that
+// adding four of those won't overflow a uint32.
+//
+// x and y are both assumed to be in the range [0, 0xffff].
+func sqDiff(x, y uint32) uint32 {
+	d := x - y
+	return (d * d) >> 2
+
+}
+
+// Find the closest color based on the chosen algorithm
+func FindClosest(pal color.Palette, orig color.Color) int {
+	if useCIE2000 {
```

Comment:
I'll show you one way to write this in Go. It will definitely feel weird in the beginning, as it uses `switch` in a strange way as compared to C. If you leave out the switch comparand, then each case is handled as a regular condition, from top to bottom.

So, the if, if-else, else would become:

```go
switch {
case useCIE2000:
	return IndexCIEDE2000(pal, orig)
case useThreshold != 0:
	return IndexMult(pal, orig, uint32(useThreshold))
default:
	return pal.Index(orig)
}
```

Use whichever way, both are equivalent. Just letting you know the `switch` version exists as well, as it is new (and a bit strange) coming from C.